### PR TITLE
Collection diffing.

### PR DESCRIPTION
**Pull request by andersio**
Depends on #33.

`diff()` is based on Paul Heckel's diffing algorithm, which is also used by IGListKit.

`diff()` can precisely detect inserts, removals, mutations and moves in O(n) time as long as all elements are unique. If there are repeated elements, `diff()` currently uniformly emits removals and inserts for every single diff until it becomes unique.

### Supported collections
`diff()` expects only collections that maintain a stable element order, e.g. `Array`. It supports:

| `Collection.Element` | Basis of comparison |
| ---- | --- |
| Any type | Custom comparator and identifier transform |
| `AnyObject` | Identified and compared by object identity |
| `AnyObject & Equatable` | Identified by object identity. Compared by value equality. Configurable. |
| `AnyObject & Hashable` | Identified by object identity. Compared by value equality. Configurable. |
| `Hashable`| Value equality |

For a glimpse in how deltas can be consumed: [the WIP `DataSource` in ReactiveCollections](https://github.com/RACCommunity/ReactiveCollections/pull/19/files#diff-03efcc672d299bd6721edd8814d8a1d4R218).

### Preconditions
1. the collection must exhibit array semantics, i.e. has a stable element order that is akin to arrays. By order the relative order is being referred to — indices may still be invalidated.

### How would the diffing algorithm react to non-unique elements?
Non-unique elements resolving to the same identity would all be emitted as delete-insert pairs until the identity again becomes unique. The order is not guaranteed to be stable for these elements.

### Why is the reference to the previous version optional?
Instantiation is not part of the `Collection` API contract. That means generically it is impossible to have "empty" collections when creating the initial snapshot.

### Checklist
- [x] Test coverage: Reproducibility.
- [x] Test coverage: Validation of primitive operation detection.
- [x] Test coverage: Validation of mixed operation sequence detection.
   _A few test cases. The reproducibility tests with random problems have coverage on this too._

### CODE REVIEW

**Comment by mdiep** [comment]
I'm on a mini-vacation, but I'll review this in a couple days when I'm back around.

**Comment by andersio** [comment]
Hmm, the algorithm for now emits correct changesets, but not a minimally viable one. Still figuring out how to deal with dependent moves (that are implied by another move with a larger step size) at a reasonable cost.

**Comment by andersio** [comment]
After a lengthy process of trial and error (check the `diffing-2` branch if you are interested in the mess), IMO it is safe to conclude that perfectly minimising the moves is costly and non-trivial. This PR for now only eliminates blocks of contiguous moves _purely_ caused by removals and insertions, which should be sufficient in common use cases.

There is still room of optimisation for simple, non overlapping moves that could be useful and should be trivial to deduce (one forward pass, one backward pass). i.e. basically the last two passes of the original Heckel's diffing algorithm.

**Comment by mdiep** [comment]
This would need a merge from `master` to be reviewed.

File: Sources/Changeset.swift [hunk 1]
```diff
@@ -116,8 +116,345 @@ public struct Changeset {
...
+			} else if isMutated {
+				mutations.insert(oldPosition)
+			}
+		}
+
+		// The following two passes perform move eliminations. The algorithm is
+		// conservative and care about only one contiguous block immediately following the
+		// deletion or the insertion.
+
+		// Pass 9: Eliminating removal-caused moves.
```
Comment (andersio) [inline Sources/Changeset.swift:274]: A fun fact about this is that I tried to adopt the existing approach used by most ObjC/Swift implementations (and previously in this branch). But when the reproducibility tests repeatedly run sufficiently large random problems with it, it starts to hit the corner cases and fails the tests.

Comment (andersio) [reply Sources/Changeset.swift:274]: Note that the move elimination algorithm at this point is not perfect — it does not know how to eliminate moves for scattered insertions and removals, but only two simple case.

But as mentioned above, it passes the reproducibility test even at larger problem size and iteration counts, so at least given the results it is presumably correct. These unideal cases can always be solved at a later point.

**Comment by andersio** [comment]
Superseded by https://github.com/RACCommunity/FlexibleDiff/pull/1.
